channels/OFF-CHAIN.md

```diff
@@ -0,0 +1,503 @@
+# Off-chain
+
+## Messages
+
+- [Overview](#overview)
+- [Control messages](#control-messages)
+	- [`error`](#error)
+	- [`ping`/`pong`](#pingpong)
+- [Establishment](#establishing-channel-off-chain)
+	- [`channel_open`](#channel_open)
+	- [`channel_accept`](#channel_accept)
+	- [`funding_created`](#funding_created)
+	- [`funding_signed`](#funding_signed)
+	- [`funding_locked`](#funding_locked)
+	- [`channel_reestablish`](#channel_reestablish)
+- Updates
+	- [`update_contract_init`](#update_contract_init)
+	- [`update_exec_func`](#update_exec_fun)
+	- [`update_exec_result`](#update_exec_result)
+	- [`update_deposit_created`](#update_deposit)
+	- [`update_deposit_signed`](#update_deposit)
+	- [`update_deposit_locked`](#update_deposit)
+	- [`update_withdrawal_created`](#update_withdrawal)
+	- [`update_withdrawal_signed`](#update_withdrawal)
+	- [`update_withdrawal_locked`](#update_withdrawal)
+- Closing
+	- [`shutdown`](#shutdown)
+	- [`closing_created`](#closing_created)
+	- [`closing_signed`](#closing_signed)
+
+
+## Overview
+
+The protocol peers use to run smart contracts is a two phase commit protocol,
+where one peer proposes a change, gets it signed off by the others and then
+commits the update locally. These locks are necessary to avoid peers getting
+confused if updates are being proposed simultaneously.
+On a higher level, to keep off chain and on chain state in sync, peers should
+refuse to sign updates for either without also getting a signature for the
+other, e.g. don't sign an on chain `channel_deposit` transaction without also
+updating the state in channel as well.
+
+With the consistency of state updates being secured between peers, it is
+essential, that peers make absolutely sure to not lose their local state, since
+it could potentially lead to them not being able to slash outdated states
+published on chain.
+
+
+## Control messages
+
```

Comment:
Actually, we should probably insert something like this:

### Framing

Each message is identified by a 1-byte message code. The size of the following message is defined by the type – see the description of each individual message type.

```
  name                  size (bytes)
 ---------------------- ----
| msg_type             | 1  |
 ---------------------- ----
| message              | .. |
 ---------------------- ----
```

The following message codes are defined:
```
  type              code
-------------------------
| channel_open     | 1  |
| channel_accept   | 2  |
| channel_reestabl | 3  |
| funding_created  | 4  |
| funding_signed   | 5  |
| funding_locked   | 6  |
| update           | 7  |
| update_ack       | 8  |
| update_error     | 9  |
| inband_message   | 96 |
| error            | 97 |
| shutdown         | 98 |
| shutdown_ack     | 99 |
-------------------------
```